src/coreclr/System.Private.CoreLib/src/System/Reflection/MethodBase.CoreCLR.cs

```diff
@@ -62,30 +65,59 @@ internal virtual Type[] GetParameterTypes()
             return parameterTypes;
         }
 
-        internal object[] CheckArguments(object[] parameters, Binder? binder,
+        private protected Span<object?> CheckArguments(ref StackAllocedArguments stackArgs, object?[]? parameters, Binder? binder,
             BindingFlags invokeAttr, CultureInfo? culture, Signature sig)
         {
-            // copy the arguments in a different array so we detach from any user changes
-            object[] copyOfParameters = new object[parameters.Length];
+            Debug.Assert(Unsafe.SizeOf<StackAllocedArguments>() == StackAllocedArguments.MaxStackAllocArgCount * Unsafe.SizeOf<object>(),
+                "MaxStackAllocArgCount not properly defined.");
 
-            ParameterInfo[]? p = null;
-            for (int i = 0; i < parameters.Length; i++)
+            Span<object?> copyOfParameters = default;
+
+            if (parameters is not null)
             {
-                object arg = parameters[i];
-                RuntimeType argRT = sig.Arguments[i];
+                // copy the arguments into a temporary buffer (or a new array) so we detach from any user changes
```

Comment:
This seems like one of those places where some enterprising developer might think "I'll use ArrayPool to ammortize this array allocation", but where even if it was beneficial in an end-to-end benchmark, it would be a case of ArrayPool enabling breaking type safety (if someone were to erroneously continue using an array they previously returned).  Might be worth a note about why the allocation here is required.